prov/shm/src/smr_progress.c

```diff
@@ -38,30 +38,45 @@
 #include "ofi_hmem.h"
 #include "ofi_atom.h"
 #include "smr.h"
+#include "smr_dsa.h"
 
 static inline void
-smr_try_progress_to_sar(struct smr_region *smr, struct smr_sar_pool *sar_pool,
-			struct smr_resp *resp, struct smr_cmd *cmd,
-			enum fi_hmem_iface iface, uint64_t device,
-			struct iovec *iov, size_t iov_count, size_t *bytes_done,
-			int *next)
+smr_try_progress_to_sar(struct smr_ep *ep, struct smr_region *smr,
+			struct smr_sar_pool *sar_pool, struct smr_resp *resp,
+			struct smr_cmd *cmd, enum fi_hmem_iface iface,
+			uint64_t device, struct iovec *iov, size_t iov_count,
+			size_t *bytes_done, int *next, void *entry_ptr)
 {
-	if (*bytes_done < cmd->msg.hdr.size)
-		smr_copy_to_sar(sar_pool, resp, cmd, iface, device, iov,
-				iov_count, bytes_done, next);
+	if (*bytes_done < cmd->msg.hdr.size) {
+		if (smr_env.use_dsa_in_sar && iface == FI_HMEM_SYSTEM) {
+			(void) smr_dsa_copy_to_sar(ep, sar_pool, resp, cmd, iov,
+					    iov_count, bytes_done, entry_ptr);
+			return;
+		} else {
+			smr_copy_to_sar(sar_pool, resp, cmd, iface, device,
+					iov, iov_count, bytes_done, next);
+		}
+	}
 	smr_signal(smr);
 }
 
 static inline void
-smr_try_progress_from_sar(struct smr_region *smr, struct smr_sar_pool *sar_pool,
-			  struct smr_resp *resp, struct smr_cmd *cmd,
-			  enum fi_hmem_iface iface, uint64_t device,
-			  struct iovec *iov, size_t iov_count,
-			  size_t *bytes_done, int *next)
+smr_try_progress_from_sar(struct smr_ep *ep, struct smr_region *smr,
+			  struct smr_sar_pool *sar_pool, struct smr_resp *resp,
+			  struct smr_cmd *cmd, enum fi_hmem_iface iface,
+			  uint64_t device, struct iovec *iov, size_t iov_count,
+			  size_t *bytes_done, int *next, void *entry_ptr)
 {
-	if (*bytes_done < cmd->msg.hdr.size)
-		smr_copy_from_sar(sar_pool, resp, cmd, iface, device, iov,
-				  iov_count, bytes_done, next);
+	if (*bytes_done < cmd->msg.hdr.size) {
+		if (smr_env.use_dsa_in_sar && iface == FI_HMEM_SYSTEM) {
+			(void) smr_dsa_copy_from_sar(ep, sar_pool, resp, cmd, iov,
+					      iov_count, bytes_done, entry_ptr);
```

Comment:
We don't need to signal here at the DSA command submission, because the signal will happen later at DSA command completion functions (dsa_update_tx_entry, dsa_update_sar_entry)